types/react-redux/index.d.ts

```diff
@@ -52,6 +53,20 @@ interface AdvancedComponentDecorator<TProps, TOwnProps> {
     (component: ComponentType<TProps>): ComponentClass<TOwnProps>;
 }
 
+/**
+ * a property P will be present if:
+ * - it is present in DecorationTargetProps
+ * - it is present in InjectedProps[P] and can satisfy DecorationTargetProps[P]
+ *   or it is not present in InjectedProps[P]
+ */
+type Matching<InjectedProps, DecorationTargetProps> = {
+	[P in keyof DecorationTargetProps]: P extends keyof InjectedProps
+		? InjectedProps[P] extends DecorationTargetProps[P]
+			? DecorationTargetProps[P]
+			: InjectedProps[P]
+		: DecorationTargetProps[P];
```

Comment:
This code looks like it is effectively doing `Pick<keyof DecorationTargetProps, InjectedProps & DecorationTargetProps>`. Is this right? If so, perhaps it should be simplified to that. If not, I would appreciate a clarification in comments on why the code does this, and how it maps to what `connect` actually does.

Reply:
That's not quite correct. The Pick version would represent something like `InjectedProps[P] & DecorationTargetProps[P]` rather than `InjectedProps[P]`. The latter is important because it's what ensures the types are matching.

Reply:
You can also modify the definition and confirm that the tests will fail with those changes.

Reply:
Okay, makes sense.